sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp

```diff
@@ -102,14 +114,180 @@ template <typename Compare = std::less<>> class default_sorter {
+  void operator()(Group g, sycl::span<T, ElementsPerWorkItem> values,
+                  Properties properties) {
+#ifdef __SYCL_DEVICE_ONLY__
+    auto range_size = g.get_local_linear_range();
+    if (scratch_size >=
+        memory_required(Group::fence_scope, g.get_local_range().size())) {
+      size_t local_id = g.get_local_linear_id();
+      auto temp = reinterpret_cast<T *>(scratch);
+
+      // to temp
+      for (std::uint32_t i = 0; i < ElementsPerWorkItem; ++i) {
+        ::new (temp + local_id * ElementsPerWorkItem + i) T(values[i]);
+      }
+
+      // sort
+      sycl::detail::merge_sort(
+          g, std::make_tuple(temp), range_size * ElementsPerWorkItem,
+          [this](auto x, auto y) {
+            return comp(std::get<0>(x), std::get<0>(y));
+          },
+          scratch + range_size * ElementsPerWorkItem * sizeof(T) + alignof(T));
+
+      // from temp
+      std::size_t shift{};
+      for (std::uint32_t i = 0; i < ElementsPerWorkItem; ++i) {
+        if constexpr (std::is_same_v<Properties, detail::is_blocked>) {
+          shift = local_id * ElementsPerWorkItem + i;
+        } else if constexpr (std::is_same_v<Properties, detail::is_striped>) {
+          shift = i * range_size + local_id;
+        }
+        values[i] = temp[shift];
+      }
+    }
+#endif
+    (void)values;
+    (void)g;
+    (void)properties;
+  }
+
+  static constexpr std::size_t memory_required(sycl::memory_scope scope,
+                                               size_t range_size) {
+    return 2 * joint_sorter<>::template memory_required<T>(scope, range_size);
   }
 };
 
+template <typename T, typename U, typename CompareT = std::less<>,
+          std::size_t ElementsPerWorkItem = 1>
+class group_key_value_sorter {
+
+  CompareT comp;
+  std::byte *scratch;
+  std::size_t scratch_size;
+
+public:
+  template <std::size_t Extent>
+  group_key_value_sorter(sycl::span<std::byte, Extent> scratch_,
+                         CompareT comp_ = {})
+      : comp(comp_), scratch(scratch_.data()), scratch_size(scratch_.size()) {}
+
+  template <typename Group>
+  std::tuple<T, U> operator()(Group g, T key, U value) {
+
+    static_assert(ElementsPerWorkItem == 1,
+                  "ElementsPerWorkItem must be equal 1");
+
+    using KeyValue = std::tuple<T, U>;
+    auto this_comp = this->comp;
+    auto comp_key_value = [this_comp](const KeyValue &lhs,
+                                      const KeyValue &rhs) {
+      return this_comp(std::get<0>(lhs).front(), std::get<0>(rhs).front());
+    };
+    return group_sorter<KeyValue, ElementsPerWorkItem,
+                        decltype(comp_key_value)>(
+        sycl::span{scratch, scratch_size},
+        comp_key_value)(g, KeyValue(key, value));
+  }
+
+  template <typename Group, typename Properties>
+  void operator()(Group g, sycl::span<T, ElementsPerWorkItem> keys,
+                  sycl::span<U, ElementsPerWorkItem> values) {
+    // Properties property) {
```

Comment:
Could you please clarify why `property` is not a parameter here? You support it in this function

Reply:
It has been changed and reverted while fixing build errors.